### ISSUE TITLE
OpenAPI 3.1: visitors after the dereference visitor don't get mutated elements

### ISSUE THREAD

**Issue by glowcloud** [swagger-api/swagger-js#3520]
### Content & configuration

Swagger/OpenAPI definition:
```yaml
openapi: 3.1.0
info:
  title: Example generation test
  version: 1.0.0
paths:
  /foo:
    get:
      responses:
        '200':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Foo'
components:
  schemas:
    Foo:
      type: object
      properties:
        bar:
          allOf:
            - type: object
              properties:
                a:
                  type: integer
```

### Describe the bug you're encountering

The `allOf` properties are not merged into the parent schema if the schema was referenced. The dereferenced schema in the response will still have the `allOf` keyword:

<img width="297" alt="Screenshot 2024-05-16 at 10 01 44" src="https://github.com/swagger-api/swagger-js/assets/44094416/07651f63-9517-4bbf-b8e0-1e74fc9d7c7d">

while the schema itself will be merged:

<img width="221" alt="Screenshot 2024-05-16 at 09 59 45" src="https://github.com/swagger-api/swagger-js/assets/44094416/a34bec39-e9cb-4407-894a-944dd07af9b6">

The issue here is that the parent elements get mutated by the `DereferenceVisitor`:
https://github.com/swagger-api/swagger-js/blob/fa77bfd689e070c2e1fe7f6f58a7cd5b9368ebea/src/resolver/apidom/reference/dereference/strategies/openapi-3-1-swagger-client/visitors/dereference.js#L776-L780
and the subsequent visitors do not see this change. 

We get the correct result if instead of merging the `AllOfVisitor` with the `DereferenceVisitor` and visiting here:
https://github.com/swagger-api/swagger-js/blob/fa77bfd689e070c2e1fe7f6f58a7cd5b9368ebea/src/resolver/apidom/reference/dereference/strategies/openapi-3-1-swagger-client/index.js#L107-L118

we did a second merge and visit with only the `AllOfVisitor` on the dereferenced element:

```js
  if (this.mode !== 'strict') {
    const allOfVisitor = AllOfVisitor({ options });
    secondVisitors.push(allOfVisitor);
   }

   // establish root visitor by visitor merging
   const rootVisitor = mergeAllVisitorsAsync(visitors, { nodeTypeGetter: getNodeType });

   let dereferencedElement = await visitAsync(refSet.rootRef.value, rootVisitor, {
    keyMap,
    nodeTypeGetter: getNodeType,
   });

   const secondRootVisitor = mergeAllVisitorsAsync(secondVisitors, {
    nodeTypeGetter: getNodeType,
   });

   dereferencedElement = await visitAsync(dereferencedElement, secondRootVisitor, {
    keyMap,
    nodeTypeGetter: getNodeType,
   });
```

because we get updated elements.

### Expected behavior

The updated elements need to be passed to the visitors after dereferencing, as this issue will affect all of them.

**Comment by char0n** [comment]
I've researched standardized approach to the problem we currently have - it's called `requeueing the traversal` and I'm currently in the process to implement it in ApiDOM. There are two approaches to the requeueing:

1. Automatic
2. Manual

Babel is using automatic requeueing as it has an abstraction facilitated via `NodePath`. 

Our traversal mechanism is more generic and not bound to particular structure (it handles parser CSTs, ASTs and ApiDOM). It was originally build as immutable, but for dereferencing we needed to utilize mutations. Every time the mutation is introduced that significantly modifies the ApiDOM tree (e.g. replacing current node), we need to manually dispatch `requeueing traversal`.
